test/test_HP_Optimization.py

```diff
@@ -0,0 +1,110 @@
+		simple_space = [["interval", 5, 3.3], [7], ["interval_int", 4.2, 1.1], ["interval", 7]]
+		cmp_space = [["interval", 3.3, 5], ["list", 1], ["interval_int", 1, 4], ["interval", 7, 7]]
+		HPO_simple = hpo.HP_Optimization(simple_function, simple_space)
+		self.assertEqual(HPO_simple.hp_space, cmp_space)
+		simple_space = [["list", 0.42, 3.3], ["interval", 2, 4], ["interval_int", 2, 4]]
+		HPO_simple = hpo.HP_Optimization(simple_function, simple_space)
+		self.assertEqual(HPO_simple.hp_space, [["list", 0.42, 3.3], ["interval", 2, 4], ["interval_int", 2, 4]])
+		#test cart_prod_hp_space
+		cart_prod = HPO_simple.cart_prod_hp_space(1)
+		cmp_cart_prod = [(0.42, 2.0, 2), (0.42, 2.0, 4), (0.42, 4.0, 2), (0.42, 4.0, 4), (3.3, 2.0, 2), (3.3, 2.0, 4), (3.3, 4.0, 2), (3.3, 4.0, 4)]
+		self.assertEqual(cart_prod, cmp_cart_prod)
+		#test perform_eval_at
+		res = HPO_simple.perform_evaluation_at([0.42, 4.0, 2])
+		self.assertEqual(res, 4.42)
+		res2 = HPO_simple.perform_evaluation_at([42, 1337])
+		self.assertEqual(res2, 1379)
+		#test GO
+		simple_space = [["list", 0.42, 1, 2.5, 3.3], ["list", 2, 4]]
+		HPO_simple = hpo.HP_Optimization(simple_function, simple_space)
+		res_GO = HPO_simple.perform_GO()
+		np.testing.assert_almost_equal(res_GO[0], [3.3, 4])
+		self.assertAlmostEqual(res_GO[1], 7.3)
+		##test RO - should I even test RO?? It's usually not that close to the optimum
+		#message = "The optimum found by perform_RO is not close enough to the actual optimum"
+		#res_RO = HPO_simple.perform_RO(15)
+		#np.testing.assert_almost_equal(res_RO[0], [3.3, 4], 1, message)
+		#self.assertAlmostEqual(res_RO[1], 7.3, 2)
+
+		#test round x
+		simple_space = [["list", 0.42, 1, 2.5, 3.3], ["interval", 2, 4], ["interval_int", 1, 5]]
+		HPO_simple = hpo.HP_Optimization(simple_function, simple_space)
+		x = [0.4, 3.33, 2.7]
+		x_rd = HPO_simple.round_x(x)
+		y = [2.5, 5, 5]
+		y_rd = HPO_simple.round_x(y)
+		self.assertEqual(x_rd, [0.42, 3.33, 3])
+		self.assertEqual(y_rd, [2.5, 4, 5])
+		#test get random x
+		x = HPO_simple.create_random_x()
+		x_rd = HPO_simple.round_x(x)
+		self.assertEqual(x, x_rd)
+		#test create_evidence_set
+		C = HPO_simple.create_evidence_set(5, 3)
+		self.assertEqual(len(C[0]), 9)
+		self.assertEqual(len(C[0][0]), 3)
+		self.assertEqual(HPO_simple.perform_evaluation_at(C[0][0]), C[1][0])
+		cmp = [0, 0, 0]
+		np.testing.assert_equal(C[0][7], cmp)
+		#test get_bounds_and_x0
+		#bounds are supposed to be smallest and biggest possible values, x0 consists of smallest values
+		bounds_and_x0 = HPO_simple.get_bounds_and_x0(0)
+		self.assertEqual(bounds_and_x0[0][0][0], 0.42)
+		self.assertEqual(bounds_and_x0[0][0][1], 3.3)
+		self.assertEqual(bounds_and_x0[0][1][0], 2)
+		self.assertEqual(bounds_and_x0[0][1][1], 4)
+		self.assertEqual(bounds_and_x0[0][2][0], 1)
+		self.assertEqual(bounds_and_x0[0][2][1], 5)
+
+		self.assertEqual(bounds_and_x0[1][0], 0.42)
+		self.assertEqual(bounds_and_x0[1][1], 2)
+		self.assertEqual(bounds_and_x0[1][2], 1)
```

Comment:
Maybe split up in several smaller test cases (this makes it easier to locate the error)